pkg/distributor/distributor.go

```diff
@@ -858,6 +859,100 @@ func (d *Distributor) PushWithCleanup(ctx context.Context, req *mimirpb.WriteReq
 	return &mimirpb.WriteResponse{}, firstPartialErr
 }
 
+// StartBackfill requests the starting of a backfill session.
+func (d *Distributor) StartBackfill(ctx context.Context, tenantID int, blockID string) error {
+	level.Info(d.log).Log("msg", "starting backfill", "tenantId", tenantID, "blockId", blockID)
+	// TODO: Verify that block hasn't already been ingested
+	return nil
+}
+
+// AddBackfillFile adds a file to an ongoing metrics backfill.
+func (d *Distributor) AddBackfillFile(ctx context.Context, tenantID int, blockID, pth string, r *http.Request) error {
+	tenantIDFromCtx, err := tenant.TenantID(ctx)
+	if err != nil {
+		return err
+	}
+	localCtx := user.InjectOrgID(ctx, tenantIDFromCtx)
+	// Get clientIP(s) from Context and add it to localCtx
+	source := util.GetSourceIPsFromOutgoingCtx(ctx)
+	localCtx = util.AddSourceIPsToOutgoingContext(localCtx, source)
+	if sp := opentracing.SpanFromContext(ctx); sp != nil {
+		localCtx = opentracing.ContextWithSpan(localCtx, sp)
+	}
+
+	level.Info(d.log).Log("msg", "adding file to metrics backfill", "tenantId", tenantID,
+		"blockId", blockID, "path", pth, "size", r.ContentLength)
+	op := ring.WriteNoExtend
+	if d.cfg.ExtendWrites {
+		op = ring.Write
+	}
+	key, err := d.tokenForLabels(tenantIDFromCtx, nil)
+	if err != nil {
+		return err
+	}
+	subRing := d.ingestersRing.ShuffleShard(tenantIDFromCtx, d.limits.IngestionTenantShardSize(tenantIDFromCtx))
+	if err := ring.DoBatch(ctx, op, subRing, []uint32{key}, func(ingester ring.InstanceDesc, indexes []int) error {
+		h, err := d.ingesterPool.GetClientFor(ingester.Addr)
+		if err != nil {
+			return err
+		}
+		c := h.(ingester_client.IngesterClient)
+
+		stream, err := c.AddBackfillFile(localCtx)
+		if err != nil {
+			return errors.Wrap(err, "failed to get gRPC stream for adding file to backfill")
+		}
+
+		buf := make([]byte, 8192)
+		var bytesWritten int64
+		for {
+			n, err := r.Body.Read(buf)
+			if err != nil && (err != io.EOF || n == 0) {
+				if err == io.EOF {
+					break
+				}
+
+				return errors.Wrap(err, "failed to read HTTP request body")
```

Comment:
Everytime we return from the function, interrupting the stream of chunks from the distributor to the ingester, what does happen in the ingester? Does the ingester correctly handle the interruption? Can you test it?

Reply:
Haven't given this any thought, I can look at it when things are in place and I can refine.